lib/optimizer/src/Transformer/ServerSideRendering.php

```diff
@@ -620,4 +699,260 @@ private function isAmpExperimentUsed(DOMElement $element)
+            }
+        }
+
+        $this->ampCustomStyleElement->textContent .= $customCss;
+        $this->ampCustomCssByteCount              += strlen($customCss);
+
+        return true;
+    }
+
+    /**
+     * Extract the custom CSS styling from a 'sizes' attribute.
+     *
+     * __sizes__
+     * "One or more strings separated by commas, indicating a set of source sizes. Each source size consists of:
+     * - A media condition. This must be omitted for the last item in the list.
+     * - A source size value."
+     *
+     * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-sizes
+     *
+     * @param Document   $document       Document containing the element to adapt.
+     * @param DOMElement $element        Element to adapt.
+     * @param string     $attributeValue Value of the attribute to extract the CSS styling from.
+     * @return string Extract custom CSS styling.
+     */
+    private function extractSizesAttributeCss(Document $document, DOMElement $element, $attributeValue)
+    {
+        if (!$element->hasAttribute(Attribute::SRCSET) || empty($element->getAttribute(Attribute::SRCSET))) {
+            // According to the Mozilla docs, a sizes attribute without a valid srcset attribute should have no effect.
+            // Therefore, it should simply be stripped, without producing media queries.
+            // @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-sizes
+            return '';
+        }
+
+        return $this->extractAttributeCss(
+            $document,
+            $element,
+            $attributeValue,
+            '#__ID__{width:%s};',
+            '@media %s{#__ID__{width:%s;}}'
+        );
+    }
+
+    /**
+     * Extract the custom CSS styling from a 'heights' attribute.
+     *
+     * __heights__ (AMP-specific)
+     * "The value of this attribute is a sizes expression based on media expressions, similar to the sizes attribute on
+     * img tags but with two key differences:
+     * - The value applies to the height, not the width of the element.
+     * - Percent values are allowed. A percent value indicates the percent of the element's width. For example, a value
+     *   of 80% indicates that the height of the element will be 80% of the element's width."
+     *
+     * @see https://amp.dev/documentation/guides-and-tutorials/learn/common_attributes/#heights
+     *
+     * @param Document   $document       Document containing the element to adapt.
+     * @param DOMElement $element        Element to adapt.
+     * @param string     $attributeValue Value of the attribute to extract the CSS styling from.
+     * @return string Extract custom CSS styling.
+     */
+    private function extractHeightsAttributeCss(Document $document, DOMElement $element, $attributeValue)
+    {
+        return $this->extractAttributeCss(
+            $document,
+            $element,
+            $attributeValue,
+            '#__ID__>i-amphtml-sizer{height:%s};',
```

Comment:
This selector is going to be problematic as it breaks AMP validation (see [spec](https://github.com/ampproject/amphtml/blob/cb57fe236c0e904b2030efcf611332d4c83fcea3/validator/validator-css.protoascii#L542-L547)):

![image](https://user-images.githubusercontent.com/134745/81238143-21ac9600-8fb6-11ea-86f5-9b23bed31f99.png)

We may have to omit extraction of `heights` for the time being.

cc @sebastianbenz

Reply:
Is it worth it to experiment with different selectors that assume a certain hierarchy without using this specific class name?

I was thinking about adding the ID directly to the sizer in this case, instead of the preceding element.
So we'd end up with this instead:
```html
'#__ID__{height:%s}'
```

I can't see why adding an ID to the sizer would cause a problem, so I'll just try it.

Reply:
> I can't see why adding an ID to the sizer would cause a problem, so I'll just try it.

The validator does not allow `id` on sizer elements.

Reply:
Are classes allowed then? Or anything else I could target with a selector?

Reply:
I'm always inserting the sizer as the first child, so I could potentially rely on `#__ID__:first-child{height:%s}` instead.

However, that will not work as expected on already optimized code where this assumption is wrong. Not sure how probably that is, though. I think to remember that all implementations insert the sizer as the first child...

Reply:
Here's the spec: https://github.com/ampproject/amphtml/blob/cb57fe236c0e904b2030efcf611332d4c83fcea3/validator/validator-main.protoascii#L4575-L4588

As I understand, the only allowed attribute is `class` and it must be `i-amphtml-sizer` which can't be used in a selector either.

Using `#ID>:first-child` would work as long as the transformer makes sure the `i-amphtml-sizer` is the first element, before any singling `[fallback]`, `[placeholder]`, or `noscript` elements.

Reply:
The `[fallback]`, `[placeholder]` and `noscript` elements would be provided outside of the Optimizer already before the markup even hits the Optimizer.

The Optimizer then does this, so it should always be true:
```php
if ($sizer) {
    $element->insertBefore($sizer, $element->firstChild);
}
```

Reply:
It seems to work, so I've adapted the code to use that for now.
Would it make sense to flag this in `amphtml`? Or are we already deep in "hacking" territory here?

I'm not sure what the long-term goal is for these attributes with regards to SSR.

Reply:
I believe I've opened the relevant issue for flagging this: https://github.com/ampproject/amphtml/issues/27468